### DISCUSSION TITLE
Distinguish between dependency resolution and scope

### DISCUSSION THREAD

**Discussion by laeubi** [eclipse-m2e/m2e-core#456] [Ideas]
Currently the dependency scope is used in the target location as follows:

- if none (or empty) is give no dependencies are resolved
- if given, dependencies are resolved and filtered for the given scope (at least in the latest version)
- if a dependency is a pom dependency and no scope is given only direct dependencies are consider otherwise all  dependencies

While implementing that feature in tycho and trying to get both in sync I noticed that this is maybe not enough and it was at other places noticed by @HannesWell that it might not fit the expectation or confuse users.

I therefore like to propose the following adjustments that would better fit the maven model and the actual dependency resolution process:

- the scope is only used to filter the (possibly empty) list of dependencies, if it is empty or not given it is assumed to be compile scope (that's how maven actually handle it)
- add a new `includeDependencyDepth` that could take the value `NONE` (no dependencies included), `DIRECT` (only direct dependency child's are included), `INFINITE` (all transitive dependencies are included)
- for pom dependencies `NONE` is consider the same as `DIRECT`

The good thing is that we can simply assume that if `includeDependencyDepth` is missing and `includeDependencyScope` is specified we can assume `INFINITE` and thus would be able to be backward compatible to the existing format with auto-upgrading from the previous format.

FYI @vogella 

Comments are welcome, if there are no concerns I'd like to get this feature into the upcoming 2.6.x release of tycho and prepare a patch for m2e as well.

**Comment by vogella** [comment]
Current approach for the Eclipse platform is described in https://bugs.eclipse.org/bugs/show_bug.cgi?id=510186. Also see https://bugs.eclipse.org/bugs/show_bug.cgi?id=570685 for the report that the SWT artifacts seems not working if sourced from Maven central.

**Reply by laeubi** [reply]
@vogella was this intended for another discussion?

**Comment by HannesWell** [comment]
That's a good proposal which would make the dependency-inclusion more powerful and more clear. So +1 from me.

In case of absent `includeDependencyScope` the automatic conversion would change it the properties to
 `includeDependencyScope = Compile`/_absent_ and `includeDependencyDepth = NONE`?

Or would the case of absent `includeDependencyScope` and `includeDependencyDepth` still be allowed and just be interpreted as `includeDependencyDepth = NONE`? I think having explicit values (at least for the depth) would be more clear.
At least we cannot mimic the Maven behaviour here, where the default that is is used in case of absence is `scope=compile` and `depth=infinite` because that would not be backward compatible.

**Reply by laeubi** [reply]
Missing values will be interpreted as includeDependencyDepth = NONE and includeDependencyScope = Compile
